tools/pkgchk/cmd/root.go

```diff
@@ -0,0 +1,254 @@
+	if !filepath.IsAbs(rootDir) {
+		asAbs, err := filepath.Abs(rootDir)
+		if err != nil {
+			return errors.Wrap(err, "failed to get absolute path")
+		}
+		rootDir = asAbs
+	}
+
+	pkgs, err := getPkgs(rootDir)
+	if err != nil {
+		return errors.Wrap(err, "failed to get packages")
+	}
+
+	var exceptions []string
+	if excepFileFlag != "" {
+		exceptions, err = loadExceptions(excepFileFlag)
+		if err != nil {
+			return errors.Wrap(err, "failed to load exceptions")
+		}
+	}
+	verifiers := getVerifiers()
+	count := 0
+	for _, pkg := range pkgs {
+		for _, v := range verifiers {
+			if err = v(pkg); err != nil && !contains(exceptions, err.Error()) {
+				fmt.Fprintln(os.Stderr, err)
+				count++
+			}
+		}
+	}
+
+	var res error
+	if count > 0 {
+		res = fmt.Errorf("found %d errors", count)
+	}
+	return res
+}
+
+func contains(items []string, item string) bool {
+	if items == nil {
+		return false
+	}
+	for _, i := range items {
+		if i == item {
+			return true
+		}
+	}
+	return false
+}
+
+func loadExceptions(excepFile string) ([]string, error) {
+	f, err := os.Open(excepFile)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	exceps := []string{}
+	for scanner := bufio.NewScanner(f); scanner.Scan(); {
+		exceps = append(exceps, scanner.Text())
+	}
+	return exceps, nil
+}
+
+type pkg struct {
+	// the directory where the package resides relative to the root dir
+	d string
+
+	// the AST of the package
+	p *ast.Package
+}
+
+// returns true if the package directory corresponds to an ARM package
+func (p pkg) isARMPkg() bool {
+	return strings.Index(p.d, "/mgmt/") > -1
+}
+
+// walks the directory hierarchy from the specified root returning a slice of all the packages found
+func getPkgs(rootDir string) ([]pkg, error) {
+	pkgs := []pkg{}
+	err := filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			// check if leaf dir
+			fi, err := ioutil.ReadDir(path)
+			if err != nil {
+				return err
+			}
+			hasSubDirs := false
+			for _, f := range fi {
+				if f.IsDir() {
+					hasSubDirs = true
+					break
+				}
+			}
+			if !hasSubDirs {
+				fs := token.NewFileSet()
+				packages, err := parser.ParseDir(fs, path, nil, parser.PackageClauseOnly)
+				if err != nil {
+					return err
+				}
+				if len(packages) < 1 {
+					return errors.New("didn't find any packages which is unexpected")
+				}
+				if len(packages) > 1 {
+					return errors.New("found more than one package which is unexpected")
+				}
+				var p *ast.Package
+				for _, pkgs := range packages {
+					p = pkgs
+				}
+				// normalize directory separator to '/' character
+				pkgs = append(pkgs, pkg{
+					d: strings.Replace(path[len(rootDir):], "\\", "/", -1),
+					p: p,
+				})
+			}
+		}
+		return nil
+	})
+	return pkgs, err
+}
+
+type verifier func(p pkg) error
+
+// returns a list of verifiers to execute
+func getVerifiers() []verifier {
+	return []verifier{
+		verifyPkgName(),
```

Comment:
can't we have functions directly here without returning them from another function ? it seems that the wrapper function doesn't much except returning the actual verifier

Reply:
Good catch, this is a remnant from an earlier iteration I forgot to refactor.